### PR TITLE
[rllib] [RFC] Dynamic definition of loss functions and modularization support

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Currently, defining a policy graph in TF is quite burdensome due to the need to define placeholders and wire them up manually. However, the dtype and shapes of these placeholders can be inferred from the postprocessed trajectory batches if we delay the construction of the loss function.

This PR:
- Adds `DynamicTFPolicyGraph`, which implements the above delayed loss function construction which allows the loss to be defined "eagerly".
- Adds `build_tf_policy` function for building dynamic TF policy graphs from modular components.
- Adds `build_trainer` function for building trainers from modular components.
- Ports the simple policy gradient example and PPO to use the dynamic graph build / builder pattern.
- Also implement the same pattern for PyTorch algos with `build_torch_policy`.

## Related issue number

This partially solves https://github.com/ray-project/ray/issues/4788

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/rllib/agents/pg/pg.py [hunk 1]
```diff
@@ -22,40 +21,24 @@
...
+            PGTorchPolicyGraph
+        return PGTorchPolicyGraph
+    else:
+        return PGPolicyGraph
+
+
+PGTrainer = build_trainer(
+    "PG",
+    default_config=DEFAULT_CONFIG,
+    default_policy_graph=PGPolicyGraph,
```
Comment (richardliaw) [inline python/ray/rllib/agents/pg/pg.py]: why do you still need DEFAULT Policy graph if you already have the creator below?

File: python/ray/rllib/agents/pg/pg.py [hunk 2]
```diff
@@ -22,40 +21,24 @@
...
+    get_policy_graph=_get_policy_graph,
+    make_policy_optimizer=_make_policy_optimizer)
```
Comment (richardliaw) [inline python/ray/rllib/agents/pg/pg.py]: `make` -> `creator`?

File: python/ray/rllib/agents/pg/pg.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/agents/pg/pg.py]: Just for consistency with existing algorithms that assign _policy_graph. Also, it is optional.

**Comment by ericl** [comment]
Update: now the loss is constructed using a dummy batch instead of the first real sample batch. This simplifies things quite a bit since the graph is guaranteed to be built when it's constructor returns.

**Review by bbalaji-ucsd — COMMENTED** [review]
If you are switching to eager execution with dynamic graphs, why not use PyTorch or TF2.0 which use eager by default?

From the current builder implementation, I think the model and exploration choices should be visible and easily swappable. I understand some of these are in config now, but they can be difficult to find and override.

**Comment by ericl** [comment]
> If you are switching to eager execution with dynamic graphs, why not use PyTorch or TF2.0 which use eager by default?

I'd like to enable eager execution by default with TF2. For compatibility reasons we will have to support TF1.x for a long time though, so ideally all code could be written to work in both modes.

> From the current builder implementation, I think the model and exploration choices should be visible and easily swappable. I understand some of these are in config now, but they can be difficult to find and override.

For the model, would something like providing your own `model_builder(input_dict, action_space, obs_space, logit_dim, config, state_in, seq_lens) -> (action_sampler: Tensor, action_prob: Tensor)` work? This would allow arbitrary models / action distribution outputs (actually, DDPG / DQN would probably be implemented using this function).

> exploration

I'm not sure what you mean for the context of PPO, could you provide an example?

File: python/ray/rllib/agents/pg/pg.py (hunk 2 above)
Comment (ericl) [reply python/ray/rllib/agents/pg/pg.py]: Currently thinking of the following convention:

`get_foo` for trivial things returning the right class or config
`foo_fn` for functions that operate on tensors
`make_foo` for functions that return a heavyweight object
`{before,after}_foo` for callbacks

**Comment by bbalaji-ucsd** [comment]
> I'd like to enable eager execution by default with TF2. For compatibility reasons we will have to support TF1.x for a long time though, so ideally all code could be written to work in both modes.
> 

Is there a document that lists the backward compatibility requirements of RLlib? Would like to understand them in more detail.

> 
> For the model, would something like providing your own `model_builder(input_dict, action_space, obs_space, logit_dim, config, state_in, seq_lens) -> (action_sampler: Tensor, action_prob: Tensor)` work? This would allow arbitrary models / action distribution outputs (actually, DDPG / DQN would probably be implemented using this function).
> 

From the user perspective, I need to know which model is used and how I can change that. I think the above proposal suffices. 

> > exploration
> 
> I'm not sure what you mean for the context of PPO, could you provide an example?

At the basic level, I would like to know if exploration is on or off (for evaluation). In general, the user can swap the current categorical sampling with epsilon-greedy sampling or something more advanced such as curiosity based learning, random network distillation.

**Comment by ericl** [comment]
> Is there a document that lists the backward compatibility requirements of RLlib? Would like to understand them in more detail.

We have documentation on API stability here: https://ray.readthedocs.io/en/latest/rllib-dev.html#api-stability
If possible, we should try to match the documented expectations here.

For TF compatibility, the concern with mandating e.g., TF2 only is that in some organizations they run with some old (customized) TF versions and take a while to catch up to bleeding edge. Also, it would be a huge effort to port RLlib to TF2 all at once :)

> From the user perspective, I need to know which model is used and how I can change that. I think the above proposal suffices.

Sounds good.

> At the basic level, I would like to know if exploration is on or off (for evaluation). In general, the user can swap the current categorical sampling with epsilon-greedy sampling or something more advanced such as curiosity based learning, random network distillation.

So exploration is some trainable component of the policy that can modulate the action distribution (e.g., by injecting epsilon greedy noise), and transform the rewards seen by the agent?

File: python/ray/rllib/tests/test_multi_agent_env.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 import ray
 from ray.rllib.agents.pg import PGTrainer
-from ray.rllib.agents.pg.pg_policy_graph import PGPolicyGraph
+from ray.rllib.agents.pg.pg_policy_graph import PGTFPolicy
 from ray.rllib.agents.dqn.dqn_policy_graph import DQNPolicyGraph
```
Comment (richardliaw) [inline python/ray/rllib/tests/test_multi_agent_env.py]: dqnpolicy

File: python/ray/rllib/agents/ppo/appo_policy_graph.py [hunk 4]
```diff
@@ -163,333 +162,234 @@ def __init__(self,
...
-                existing_state_in = existing_inputs[7:-1]
-                existing_seq_lens = existing_inputs[-1]
-            else:
-                actions, dones, behaviour_logits, rewards, observations, \
-                    prev_actions, prev_rewards, adv_ph, value_targets = \
-                    existing_inputs[:9]
-                existing_state_in = existing_inputs[9:-1]
-                existing_seq_lens = existing_inputs[-1]
+def _make_time_major(policy, tensor, drop_last=False):
+    """Swaps batch and trajectory axis.
```
Comment (richardliaw) [inline python/ray/rllib/agents/ppo/appo_policy_graph.py]: space

File: python/ray/rllib/evaluation/tf_policy_template.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            a dict of TF fetches given the policy graph object
+        postprocess_fn (func): optional experience postprocessing function
+            that takes the same args as PolicyGraph.postprocess_trajectory()
+        optimizer_fn (func): optional function that returns a tf.Optimizer
+            given the policy graph object
+        gradients_fn (func): optional function that returns a list of gradients
+            given a tf optimizer and loss tensor. If not specified, this
+            defaults to optimizer.compute_gradients(loss)
+        before_init (func): optional function to run at the beginning of
+            __init__ that takes the same arguments as __init__
```
Comment (richardliaw) [inline python/ray/rllib/evaluation/tf_policy_template.py]: __init__ is confusing here

**Review by richardliaw — APPROVED** [review]
did not look at `dynamic_tf_policy_graph` too carefully otherwise as an initial implementation refactor looks good

File: python/ray/rllib/evaluation/tf_policy_template.py (hunk 5 above)
Comment (richardliaw) [reply python/ray/rllib/evaluation/tf_policy_template.py]: basically, I don't know which __init__ this is referring
